document.go

```diff
@@ -26,16 +26,42 @@ type Document struct {
 }
 
 // Add a VerificationMethod as AssertionMethod
+// If the controller is not set, it will be set to the documents ID
 func (d *Document) AddAssertionMethod(v *VerificationMethod) {
-	if v.Controller.Empty() {
-		v.Controller = d.ID
-	}
-	d.VerificationMethod = append(d.VerificationMethod, v)
+	d.addVerificationMethodIfNotExists(v)
 	d.AssertionMethod = append(d.AssertionMethod, VerificationRelationship{
 		VerificationMethod: v,
 		reference:          v.ID,
 	})
+}
 
+// AddAuthenticationMethod adds a VerificationMethod as AuthenticationMethod
+// If the controller is not set, it will be set to the documents ID
+func (d *Document) AddAuthenticationMethod(v *VerificationMethod) {
+	d.addVerificationMethodIfNotExists(v)
+	d.Authentication = append(d.Authentication, VerificationRelationship{
+		VerificationMethod: v,
+		reference:          v.ID,
+	})
+}
+
+// When adding verificationMethods, this metho makes sure there won't be any duplicates
+func (d *Document) addVerificationMethodIfNotExists(v*VerificationMethod) {
+	for _, ptr := range d.VerificationMethod {
+		// check if the pointer is already in the list
```

Comment:
I think you can just compare the actual structs;

```
func TestStructEqualness(t *testing.T) {
	type someValue struct {
		value string
	}

	assert.Equal(t, someValue{"foo"}, someValue{"foo"})
	assert.True(t, someValue{"foo"} == someValue{"foo"})
	assert.NotEqual(t, someValue{"foo"}, someValue{"bar"})
	assert.False(t, someValue{"foo"} == someValue{"bar"})
}
```

On a second thought, comparing the IDs is functionally more correct, probably (since you can't have 2 with the same ID).

Didn't think of it earlier, but does case sensitivy matter?

Reply:
Yes, it compares strings. I added a note to the godoc DID.Equal() method.